src/common/walFilter/walFilter.c

```diff
@@ -0,0 +1,763 @@
+        THROW(FormatError, "no WAL files were found in the repository");
+    }
+
+    uint64_t segnoDiff = UINT64_MAX;
+    for (uint32_t i = 0; i < strLstSize(segmentList); i++)
+    {
+        const String *file = strSubN(strLstGet(segmentList, i), 0, 24);
+        TimeLineID tli;
+        uint64_t fileSegNo = 0;
+        XLogFromFileName(strZ(file), &tli, &fileSegNo, this->segSize);
+
+        if (fileSegNo - segno < segnoDiff && fileSegNo > segno)
+        {
+            segnoDiff = fileSegNo - segno;
+            walSegment = strLstGet(segmentList, i);
+        }
+    }
+
+    // current file is newest
+    if (segnoDiff == UINT64_MAX)
+    {
+        memContextSwitchBack();
+        memContextDiscard();
+        return false;
+    }
+
+    const StorageRead *const storageRead = storageNewReadP(
+        storageRepoIdx(0),
+        strNewFmt(STORAGE_REPO_ARCHIVE "/%s/%s", strZ(this->archiveInfo->archiveId), strZ(walSegment)));
+
+    buildArchiveGetPipeLine(ioReadFilterGroup(storageReadIo(storageRead)), this->archiveInfo);
+
+    ioReadOpen(storageReadIo(storageRead));
+
+    Buffer *const buffer = bufNew(this->walPageSize);
+    size_t size = ioRead(storageReadIo(storageRead), buffer);
+    bufUsedSet(buffer, size);
+    while (this->got_len != this->record->xl_tot_len)
+    {
+        if (!readRecord(this, buffer))
+        {
+            if (ioReadEof(storageReadIo(storageRead)))
+            {
+                THROW(FormatError, "Unexpected WAL end");
+            }
+
+            bufUsedZero(buffer);
+            size = ioRead(storageReadIo(storageRead), buffer);
+            bufUsedSet(buffer, size);
+        }
+    }
+
+    MEM_CONTEXT_TEMP_END();
+    return true;
+}
+
+static void
+walFilterProcess(THIS_VOID, const Buffer *const input, Buffer *const output)
+{
+    THIS(WalFilterState);
+
+    FUNCTION_LOG_BEGIN(logLevelTrace);
+        FUNCTION_LOG_PARAM(WAL_FILTER, this);
+        FUNCTION_LOG_PARAM(BUFFER, input);
+        FUNCTION_LOG_PARAM(BUFFER, output);
+    FUNCTION_LOG_END();
+
+    ASSERT(this != NULL);
+    ASSERT(output != NULL);
+
+    // Avoid creating local variables before the record is fully read.
+    // Since if the input buffer is exhausted, we can exit the function.
+
+    if (input == NULL)
+    {
+        // We have an incomplete record at the end
+        if (this->currentStep != 0)
+        {
+            const size_t size_on_page = this->got_len;
+
+            // if xl_info and xl_rmid of the header is in current file then read end of record from next file if it exits
+            if (this->got_len >= offsetof(XLogRecord, xl_rmid) + SIZE_OF_STRUCT_MEMBER(XLogRecord, xl_rmid) && getEndOfRecord(this))
+            {
+                filterRecord(this);
+            }
+
+            checkOutputSize(output, size_on_page);
+            memcpy(bufRemainsPtr(output), this->record, size_on_page);
+            bufUsedInc(output, size_on_page);
```

Comment:
It seems that the `bufCatC` function grows the buffer if needed, copies data into it, and sets the size to use, so this function can be used here and in other similar places instead of the three.